File: ramanujan/presets.py
"""Preset store + deadlock guard — ramanujan/presets.py (v2 Phase 3).

Named presets are validated against `family_registry.json`; <2 distinct families
can never produce a panel-verified stop (Tier 2 excludes the caller's family).
Save is NEVER blocked — even a single-family preset warns explicitly and still
proceeds, per §4.2. Keys never appear in presets.json (refs only).

File: ~/.config/ramanujan/presets.json (0600 via atomic tempfile+replace).
Env override: RAMANUJAN_PRESETS (tests).
Structure: {"diverse-5": ["anthropic/claude-opus-5", ...], ...}
"""

from __future__ import annotations

import contextlib
import json
import os
import stat
import tempfile
from pathlib import Path

from ramanujan.families import DEFAULT_REGISTRY_PATH, PresetCheck, load_registry, validate_preset

DEFAULT_PRESETS_PATH = Path.home() / ".config" / "ramanujan" / "presets.json"


def _presets_path() -> Path:
    env = os.environ.get("RAMANUJAN_PRESETS")
    if env:
        return Path(env).expanduser()
    return DEFAULT_PRESETS_PATH


def load_presets(path: Path | str | None = None) -> dict[str, list[str]]:
    p = Path(path) if path is not None else _presets_path()
    if not p.exists():
        return {}
    try:
        data = json.loads(p.read_text(encoding="utf-8"))
    except Exception as e:
        raise ValueError(f"presets file invalid JSON at {p}: {e}") from e
    if not isinstance(data, dict):
        raise ValueError(f"presets must be a JSON object at {p}")
    out: dict[str, list[str]] = {}
    for k, v in data.items():
        if not isinstance(k, str) or not k.strip():
            raise ValueError(f"invalid preset name: {k!r}")
        if not isinstance(v, list) or not all(isinstance(x, str) and x.strip() for x in v):
            raise ValueError(f"preset {k!r} must be a list of non-empty strings")
        out[k.strip()] = [x.strip() for x in v]  # type: ignore[arg-type]
    return out


def save_presets(presets: dict[str, list[str]], path: Path | str | None = None) -> None:
    p = Path(path) if path is not None else _presets_path()
    p.parent.mkdir(parents=True, exist_ok=True)
    content = json.dumps(presets, indent=2, ensure_ascii=False) + "\n"
    fd, tmp = tempfile.mkstemp(dir=p.parent, prefix=p.name + ".", suffix=".tmp")
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as fh:
            fh.write(content)
        os.replace(tmp, p)
        with contextlib.suppress(Exception):
            os.chmod(p, stat.S_IRUSR | stat.S_IWUSR)
    except BaseException:
        with contextlib.suppress(OSError):
            os.unlink(tmp)
        raise


def get_preset(name: str, path: Path | str | None = None) -> list[str] | None:
    return load_presets(path).get(name)


def check_preset(name: str, models: list[str], registry_path: Path | str | None = None) -> PresetCheck:
    reg = load_registry(registry_path or DEFAULT_REGISTRY_PATH)
    return validate_preset(name, models, reg)


def validate_all_presets(presets: dict[str, list[str]], registry_path: Path | str | None = None) -> dict[str, PresetCheck]:
    reg = load_registry(registry_path or DEFAULT_REGISTRY_PATH)
    return {n: validate_preset(n, m, reg) for n, m in presets.items()}


def set_preset(
    name: str,
    models: list[str],
    path: Path | str | None = None,
    registry_path: Path | str | None = None,
) -> PresetCheck:
    """Validate (warning on deadlock) and persist. Always saves; never hard-blocks."""
    check = check_preset(name, models, registry_path)
    presets = load_presets(path)
    presets[name] = list(models)
    save_presets(presets, path)
    return check


def delete_preset(name: str, path: Path | str | None = None) -> bool:
    presets = load_presets(path)
    if name not in presets:
        return False
    del presets[name]
    save_presets(presets, path)
    return True

File: agent/packages/config/src/presets.ts
/**
 * Preset store — mirrors ramanujan/presets.py.
 * ~/.config/ramanujan/presets.json (JSON object name -> string[]), env RAMANUJAN_PRESETS.
 * Warns on deadlock (validatePreset), never blocks.
 */

import { chmodSync, existsSync, mkdirSync, readFileSync, renameSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { homedir } from "node:os";
import { loadRegistry, validatePreset, type PresetCheck } from "./families.ts";

function defaultPresetsPath(): string {
	const over = process.env["RAMANUJAN_PRESETS"];
	if (over) return over;
	return join(homedir(), ".config", "ramanujan", "presets.json");
}

export function loadPresets(path?: string): Record<string, string[]> {
	const p = path ?? defaultPresetsPath();
	if (!existsSync(p)) return {};
	const data = JSON.parse(readFileSync(p, "utf-8"));
	if (typeof data !== "object" || data === null || Array.isArray(data)) throw new Error(`presets must be object at ${p}`);
	const out: Record<string, string[]> = {};
	for (const [k, v] of Object.entries(data)) {
		if (typeof k !== "string" || !k.trim()) throw new Error(`invalid preset name: ${k}`);
		if (!Array.isArray(v) || !v.every((x) => typeof x === "string" && (x as string).trim())) throw new Error(`preset ${k} must be string[]`);
		out[k.trim()] = (v as string[]).map((x) => x.trim());
	}
	return out;
}

export function savePresets(presets: Record<string, string[]>, path?: string): void {
	const p = path ?? defaultPresetsPath();
	mkdirSync(dirname(p), { recursive: true });
	const tmp = `${p}.${process.pid}.${Date.now()}.tmp`;
	writeFileSync(tmp, JSON.stringify(presets, null, 2) + "\n", "utf-8");
	renameSync(tmp, p);
	try {
		chmodSync(p, 0o600);
	} catch {
		// best effort
	}
}

export function getPreset(name: string, path?: string): string[] | null {
	return loadPresets(path)[name] ?? null;
}

export function checkPreset(name: string, models: string[], registryPath?: string): PresetCheck {
	const reg = loadRegistry(registryPath);
	return validatePreset(name, models, reg);
}

export function setPreset(name: string, models: string[], path?: string, registryPath?: string): PresetCheck {
	const reg = loadRegistry(registryPath);
	const check = validatePreset(name, models, reg);
	const presets = loadPresets(path);
	presets[name] = [...models];
	savePresets(presets, path);
	return check;
}

export function deletePreset(name: string, path?: string): boolean {
	const presets = loadPresets(path);
	if (!(name in presets)) return false;
	delete presets[name];
	savePresets(presets, path);
	return true;
}
